torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,303 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+#include <initializer_list>
+
+namespace torch {
+
+namespace detail {
+
+enum class TensorDataContainerType { Scalar, InitList, Tensor };
+
+struct TensorDataContainer;
+
+inline std::ostream& operator<<(std::ostream& stream, const TensorDataContainer& tensor_data_container);
+
+// We use `TensorDataContainer` to support converting the following data container types
+// into the equivalent Tensor:
+//
+// 1. Arbitrarily nested braced-init-list (e.g. `{{1, 2}, {3, 4}}`).
+// 2. `at::ArrayRef` of supported tensor data types.
+// 3. `std::vector` of supported tensor data types.
+//
+// At any time, a `TensorDataContainer` object represents one of the following:
+//
+// 1. A scalar with value `scalar()` and type `scalar_type()`.
+// 2. A Tensor represented in `std::initializer_list<TensorDataContainer>` form,
+//    with value `init_list()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
+// 3. A Tensor represented in `at::Tensor` form, with value `tensor()`, scalar type `scalar_type()`,
+//    and Tensor sizes `sizes()`.
+//
+// All the infrastructure here is mostly to support converting an arbitrarily nested braced-init-list
+// to the equivalent Tensor successfully. Consider the following example:
+//
+// `torch::tensor({{1}, {2}})`
+//
+// Here is the code path that it goes through:
+//
+// `at::Tensor tensor(detail::TensorDataContainer tensor_data_container)`
+//
+// which calls:
+//
+// `TensorDataContainer({{1}, {2}})`
+//
+// which matches to the `TensorDataContainer(std::initializer_list<TensorDataContainer>)` constructor,
+// and in an attempt to convert `{1}` and `{2}` to `TensorDataContainer`, it calls the following:
+//
+// `TensorDataContainer({1})`  (same call path happens for `{2}`, and we'll just focus on `{1}` here)
+//
+// At this point, theoretically there are two plausible ways for `{1}` to be matched to one of the
+// constructors of `TensorDataContainer`:
+//
+// 1. It can be a list-initialization of a scalar value, thus matching `TensorDataContainer(int value)`.
+// 2. It can be converted to `std::initializer_list<TensorDataContainer>`, thus matching
+//    `TensorDataContainer(std::initializer_list<TensorDataContainer>)`.
+//
+// How does the compiler decide which one to choose? According to
+// `https://en.cppreference.com/w/cpp/language/list_initialization`, braced-init-list always prefers
+// the constructor that takes `std::initializer_list`. Hence we happily move forward with constructor #2,
+// and it calls the following:
+//
+// `TensorDataContainer(1)`
+//
+// Now it matches `TensorDataContainer(int value)`, which stores `1` as a scalar value. All is good.
+//
+// Note that `torch::tensor({{1}, {2}})` can also match another previously existing function overload:
```

Comment:
We don't really need this paragraph (which is a retrospective on why the old approach didn't work). Something like this is good to have in the commit message / pull request body but devs trying to understand how TensorDataConstructor works only need to know that it is used for calling `torch::tensor` with ArrayRef, vectors, and braced initialization lists.

Reply:
Removed the paragraph :D